libsolidity/ast/ASTEnums.h

```diff
@@ -50,5 +51,22 @@ inline std::string stateMutabilityToString(StateMutability const& _stateMutabili
 	}
 }
 
+class Type;
+
+/// Container for function call parameter types & names
+struct FuncCallArguments
+{
+	/// Types of arguments
+	std::shared_ptr<std::vector<std::shared_ptr<Type const>>> types;
```

Comment:
We usually do not use shared pointer to const Types - Types are non-mutable anyway.

Reply:
Also, is the outermost shared_ptr needed?

Reply:
that's not exactly true though? I mean we're using `using TypePointer = std::shared_ptr<Type const>;` everywhere which is exactly that.

I suppose we don't need the outmost. To be fair, I haven't completely figured out what the logic/policy is regarding shared_ptr in our code :) 

So Types generally don't need shared_ptr because they will always exist throughout the compilation?

Reply:
Not using `TypePointer` aka `std::shared_ptr<Type const>` here will make it more complicated to reuse that vector in other functions (e.g. `ASTJsonConverter::typePointerToJson(std::shared_ptr<std::vector<TypePointer>> _tps)`). Of course I can change `TypePointer` in all those functions as well.

Reply:
Ah sorry, then I'm totally wrong - but you could use `TypePointer`.

Reply:
I would use `TypePointer` but that is defined in `Types.h` :D